indexing-service/src/main/java/io/druid/indexing/common/task/ArchiveTask.java

```diff
@@ -93,12 +101,41 @@ public TaskStatus run(TaskToolbox toolbox) throws Exception
       log.info("OK to archive segment: %s", unusedSegment.getIdentifier());
     }
 
-    // Move segments
+
+    // Archiving segments can take time which seems to be a lot of work to do in critical section
+    // Thus, after each batch downgrade the lock to ensure liveliness of the system
+    // so that higher priority tasks are not blocked for a long time
+    // Note - If we are unable to upgrade the lock again then the task will fail,
+    // there is no point in retrying here as upgrade should fail only when the taskLock is removed by a higher priority task
+    int counter = 0;
     for (DataSegment segment : unusedSegments) {
+      if (counter % getBatchSize() == 0) {
+        // LockToggleAction is idempotent
+        if (!toolbox.getTaskActionClient().submit(new LockToggleAction(getInterval(), TaskLockToggle.DOWNGRADE))) {
+          throw new ISE(
+              "Lock downgrade failed for interval [%s] !! Successfully archived [%s] segments out of [%s] before failing",
+              getInterval(),
+              counter,
+              unusedSegments.size()
+          );
+        }
+        // Give sometime to waiting higher priority tasks (if any) to acquire the lock
+        Thread.sleep(10);
+        // Try to upgrade the lock again - we will be successful if no other task needs it
+        if (!toolbox.getTaskActionClient().submit(new LockToggleAction(getInterval(), TaskLockToggle.UPGRADE))) {
+          throw new ISE(
+              "Lock upgrade failed for interval [%s] !! Successfully archived [%s] segments out of [%s] before failing",
+              getInterval(),
+              counter,
+              unusedSegments.size()
+          );
+        }
+      }
```

Comment:
Is this logic turned inside out? are you actually trying to upgrade, submit the batch quantity of  actions, then downgrade for a short amount of time?